server/pipeline/create.go

```diff
@@ -37,109 +36,65 @@ func Create(ctx context.Context, _store store.Store, repo *model.Repo, pipeline
-	// fetch the pipeline file from the forge
-	configFetcher := forge.NewConfigFetcher(server.Config.Services.Forge, server.Config.Services.Timeout, server.Config.Services.ConfigService, repoUser, repo, pipeline)
-	forgeYamlConfigs, configFetchErr = configFetcher.Fetch(ctx)
-	if configFetchErr == nil {
-		filtered, parseErr = checkIfFiltered(repo, pipeline, forgeYamlConfigs)
-		if parseErr == nil {
-			if filtered {
-				err := ErrFiltered{Msg: "global when filter of all workflows do skip this pipeline"}
-				log.Debug().Str("repo", repo.FullName).Msgf("%v", err)
-				return nil, err
-			}
-
-			if zeroSteps(pipeline, forgeYamlConfigs) {
-				err := ErrFiltered{Msg: "step conditions yield zero runnable steps"}
-				log.Debug().Str("repo", repo.FullName).Msgf("%v", err)
-				return nil, err
-			}
-		}
-	}
+	forge.Refresh(ctx, server.Config.Services.Forge, _store, repoUser)
 
 	// update some pipeline fields
 	pipeline.RepoID = repo.ID
 	pipeline.Status = model.StatusPending
 
+	// fetch the pipeline file from the forge
+	configFetcher := forge.NewConfigFetcher(server.Config.Services.Forge, server.Config.Services.Timeout, server.Config.Services.ConfigService, repoUser, repo, pipeline)
+	forgeYamlConfigs, configFetchErr := configFetcher.Fetch(ctx)
+
 	if configFetchErr != nil {
 		log.Debug().Str("repo", repo.FullName).Err(configFetchErr).Msgf("cannot find config '%s' in '%s' with user: '%s'", repo.Config, pipeline.Ref, repoUser.Login)
-		pipeline.Started = time.Now().Unix()
-		pipeline.Finished = pipeline.Started
-		pipeline.Status = model.StatusError
-		pipeline.Error = fmt.Sprintf("pipeline definition not found in %s", repo.FullName)
-	} else if parseErr != nil {
+		return nil, persistPipelineWithErr(ctx, _store, pipeline, repo, repoUser, fmt.Sprintf("pipeline definition not found in %s", repo.FullName))
+	}
+
+	pipelineItems, parseErr := parsePipeline(_store, pipeline, repoUser, repo, forgeYamlConfigs, nil)
+	if parseErr != nil {
 		log.Debug().Str("repo", repo.FullName).Err(parseErr).Msg("failed to parse yaml")
-		pipeline.Started = time.Now().Unix()
-		pipeline.Finished = pipeline.Started
-		pipeline.Status = model.StatusError
-		pipeline.Error = fmt.Sprintf("failed to parse pipeline: %s", parseErr.Error())
-	} else {
-		setGatedState(repo, pipeline)
+		return nil, persistPipelineWithErr(ctx, _store, pipeline, repo, repoUser, fmt.Sprintf("failed to parse pipeline: %s", parseErr.Error()))
 	}
 
+	if len(pipelineItems) == 0 {
+		log.Debug().Str("repo", repo.FullName).Msg(ErrFiltered.Error())
+		return nil, ErrFiltered
+	}
+
+	setGatedState(repo, pipeline)
+
 	err = _store.CreatePipeline(pipeline)
 	if err != nil {
-		msg := fmt.Sprintf("failure to save pipeline for %s", repo.FullName)
-		log.Error().Err(err).Msg(msg)
-		return nil, fmt.Errorf(msg)
+		msg := fmt.Errorf("failure to save pipeline for %s", repo.FullName)
+		log.Error().Err(err).Msg(msg.Error())
+		return nil, msg
 	}
 
+	pipeline = setPipelineStepsOnPipeline(pipeline, pipelineItems)
```

Comment:
Should we move this line below `setGatedState(repo, pipeline)` so those two adjusments are kinda bundeled?